### ISSUE TITLE
npm install failed for jpegtran-bin, optipng-bin

### ISSUE THREAD

**Issue by idothebest4u** [npm/npm#8682]
```
 npm --version
2.7.4
node --version
v0.12.2
```

++++++++++++++++ERROR++++++++++++

```
npm install
npm WARN deprecated CSSselect@0.7.0: the module is now available as 'css-select'
npm WARN optional dep failed, continuing fsevents@0.3.6
npm WARN deprecated bignumber.js@2.0.0: critical bug fixed in v2.0.4

> jpegtran-bin@3.0.2 postinstall /opt/urbancodeagent/agent/var/work/project/MoveCommand/Build_develop/node_modules/grunt-contrib-imagemin/node_modules/imagemin/node_modules/imagemin-jpegtran/node_modules/jpegtran-bin
> node lib/install.js

  ⚠ Request to https://raw.github.com/imagemin/jpegtran-bin/3.0.2/vendor/linux/x64/jpegtran failed
  ⚠ jpegtran pre-build test failed
  ℹ compiling from source
  ✖ GotError: Request to http://downloads.sourceforge.net/project/libjpeg-turbo/1.4.0/libjpeg-turbo-1.4.0.tar.gz failed
    at ClientRequest.<anonymous> (/opt/urbancodeagent/agent/var/work/project/MoveCommand/Build_develop/node_modules/grunt-contrib-imagemin/node_modules/imagemin/node_modules/imagemin-jpegtran/node_modules/jpegtran-bin/node_modules/bin-build/node_modules/download/node_modules/got/index.js:177:7)
    at ClientRequest.g (events.js:199:16)
    at ClientRequest.emit (events.js:107:17)
    at Socket.socketErrorListener (_http_client.js:271:9)
    at Socket.emit (events.js:107:17)
    at net.js:940:16
    at process._tickCallback (node.js:355:11)
Caused By: Error: getaddrinfo ENOTFOUND downloads.sourceforge.net
    at errnoException (dns.js:44:10)
    at GetAddrInfoReqWrap.onlookup [as oncomplete] (dns.js:94:26)
npm WARN deprecated CSSwhat@0.4.7: the module is now available as 'css-what'

> phantomjs@1.9.17 install /opt/urbancodeagent/agent/var/work/project/MoveCommand/Build_develop/node_modules/karma-phantomjs-launcher/node_modules/phantomjs
> node install.js

Download already available at /tmp/phantomjs/phantomjs-1.9.8-linux-x86_64.tar.bz2
Extracting tar contents (via spawned process)
Removing /opt/urbancodeagent/agent/var/work/project/MoveCommand/Build_develop/node_modules/karma-phantomjs-launcher/node_modules/phantomjs/lib/phantom
Copying extracted folder /tmp/phantomjs/phantomjs-1.9.8-linux-x86_64.tar.bz2-extract-1435284706244/phantomjs-1.9.8-linux-x86_64 -> /opt/urbancodeagent/agent/var/work/project/MoveCommand/Build_develop/node_modules/karma-phantomjs-launcher/node_modules/phantomjs/lib/phantom
Writing location.js file
Done. Phantomjs binary available at /opt/urbancodeagent/agent/var/work/project/MoveCommand/Build_develop/node_modules/karma-phantomjs-launcher/node_modules/phantomjs/lib/phantom/bin/phantomjs

> optipng-bin@3.0.2 postinstall /opt/urbancodeagent/agent/var/work/project/MoveCommand/Build_develop/node_modules/optipng-bin
> node lib/install.js

  ⚠ Request to https://raw.github.com/imagemin/optipng-bin/3.0.2/vendor/linux/x64/optipng failed
  ⚠ optipng pre-build test failed
  ℹ compiling from source
  ✖ GotError: Request to http://downloads.sourceforge.net/project/optipng/OptiPNG/optipng-0.7.5/optipng-0.7.5.tar.gz failed
    at ClientRequest.<anonymous> (/opt/urbancodeagent/agent/var/work/project/MoveCommand/Build_develop/node_modules/optipng-bin/node_modules/bin-build/node_modules/download/node_modules/got/index.js:177:7)
    at ClientRequest.g (events.js:199:16)
    at ClientRequest.emit (events.js:107:17)
    at Socket.socketErrorListener (_http_client.js:271:9)
    at Socket.emit (events.js:107:17)
    at net.js:940:16
    at process._tickCallback (node.js:355:11)
Caused By: Error: getaddrinfo ENOTFOUND downloads.sourceforge.net
    at errnoException (dns.js:44:10)
    at GetAddrInfoReqWrap.onlookup [as oncomplete] (dns.js:94:26)

> gifsicle@3.0.1 postinstall /opt/urbancodeagent/agent/var/work/project/MoveCommand/Build_develop/node_modules/imagemin-gifsicle/node_modules/gifsicle
> node lib/install.js

  ⚠ Request to https://raw.github.com/imagemin/gifsicle-bin/3.0.1/vendor/linux/x64/gifsicle failed
  ⚠ gifsicle pre-build test failed
  ℹ compiling from source
  ✖ GotError: Request to http://www.lcdf.org/gifsicle/gifsicle-1.87.tar.gz failed
    at ClientRequest.<anonymous> (/opt/urbancodeagent/agent/var/work/project/MoveCommand/Build_develop/node_modules/imagemin-gifsicle/node_modules/gifsicle/node_modules/bin-build/node_modules/download/node_modules/got/index.js:177:7)
    at ClientRequest.g (events.js:199:16)
    at ClientRequest.emit (events.js:107:17)
    at Socket.socketErrorListener (_http_client.js:271:9)
    at Socket.emit (events.js:107:17)
    at net.js:940:16
    at process._tickCallback (node.js:355:11)
Caused By: Error: getaddrinfo ENOTFOUND www.lcdf.org
    at errnoException (dns.js:44:10)
    at GetAddrInfoReqWrap.onlookup [as oncomplete] (dns.js:94:26)
+++++++++++++++++
npm install http://downloads.sourceforge.net/project/libjpeg-turbo/1.4.0/libjpeg-turbo-1.4.0.tar.gz .
npm WARN package.json assert@1.3.0 assert is also the name of a node core module.
npm ERR! not a package /tmp/npm-61652-b4c017a1/downloads.sourceforge.net/project/libjpeg-turbo/1.4.0/libjpeg-turbo-1.4.0.tar.gz
npm ERR! Linux 2.6.32-431.29.2.el6.x86_64
npm ERR! argv "/usr/local/bin/node" "/usr/local/bin/npm" "install" "http://downloads.sourceforge.net/project/libjpeg-turbo/1.4.0/libjpeg-turbo-1.4.0.tar.gz" "."
npm ERR! node v0.12.2
npm ERR! npm  v2.7.4
npm ERR! path /tmp/npm-61652-b4c017a1/unpack-0b9533094cdb/package.json
npm ERR! code ENOENT
npm ERR! errno -2

npm ERR! enoent ENOENT, open '/tmp/npm-61652-b4c017a1/unpack-0b9533094cdb/package.json'
npm ERR! enoent This is most likely not a problem with npm itself
npm ERR! enoent and is related to npm not being able to find a file.
npm ERR! enoent

npm ERR! Please include the following file with any support request:
npm ERR!     /opt/urbancodeagent/agent/var/work/project/MoveCommand/Build_develop/npm-debug.log
```

**Comment by samualli** [comment]
Hey guys, I'm getting the exact same issue, all my devDependencies are being downloaded and installed properly except for these exact three libraries, jpegtran-bin@3.0.2, optipng-bin@3.0.2, and gifsicle@3.0.1.  Hoping for a solution or possible explanation why soon,

![image](https://cloud.githubusercontent.com/assets/8726801/8378551/67009694-1be8-11e5-8538-0902f23e3d37.png)

**Comment by othiym23** [comment]
Are either or both of you on a network with an unusual configuration? Proxies, or DNS whitelisting, being located in China, or something similar? In both cases, the problem is that the system running the install can't resolve `downloads.sourceforge.net`, which isn't anything that npm can work around.

**Comment by othiym23** [comment]
...especially as the part that's failing is an install script that's part of the packages, and not anything inside npm.

**Comment by samualli** [comment]
I am behind a proxy, but we have our npm config set, and other libraries are being downloaded properly.  The strange thing is that a month ago, these three libraries downloaded fine.  So I have a working copy on my local machine, by any new installs have issues downloading gifsicle, jpeg and opti-png.  I've resetted my proxy and updated my node and npm just in case, but still no luck with these three libs.

**Comment by idothebest4u** [comment]
We had verified the proxy logs, where it could reach to http://registry.npmjs.org, but couldn't find any packets trying to reach  http://downloads.sourceforge.net/ or http://www.lcdf.org/.
Why its giving issue with only these 3 components. Even I tried with npm 2.12.0 and JAVA 1.8 couldn't help.
Not sure what else we are missing..trying to find out with tcmpdump in linux, hoping it could provide some way to troubleshoot.

**Comment by busches** [comment]
You have `http_proxy` and `https_proxy` set, in the command line, not just for npm, as well? Those are needed by install scripts when downloading other files. e.g. `set http_proxy=http://yourproxy:port/` Causes me issues all the time on our proxy when I first forget to set that.

**Comment by idothebest4u** [comment]
yeah I did that. Even we could able to donwload those packages using wget, but using npm it still gives the error.

**Comment by stuartmemo** [comment]
Just chipping in to say I'm having the exact same issue with gifsicle, jpegtran and opti-png. I've not updated node/npm in a while - node v0.12.0 and npm 2.5.1.

```
  ⚠ Request to https://raw.github.com/imagemin/gifsicle-bin/3.0.1/vendor/osx/gifsicle failed
  ⚠ gifsicle pre-build test failed
  ℹ compiling from source
  ✖ GotError: Request to http://www.lcdf.org/gifsicle/gifsicle-1.87.tar.gz failed
```

If I can provide any more logs to help let me know.

**Comment by MattDiMu** [comment]
I'm currently experiencing the same issue, as all other packages are downloaded, I'm quite confused...

**Comment by idothebest4u** [comment]
Downgrade bin-build from 2.1.2 to 2.1.1, now everything looks good.

**Comment by stuartmemo** [comment]
@idothebest4u I don't use bin-build anywhere in my project sadly.

**Comment by stuartmemo** [comment]
@idothebest4u Sorry! You're correct. It's nested way down deep. I can't grep to save my life.

**Comment by stuartmemo** [comment]
Downgrading worked a treat, thanks. General npm question - why does installing `bin-build@2.1.1` at the root of my project affect modules that have that as their own dependency? Should they not just use their own dependencies?

**Comment by MattDiMu** [comment]
Unfortunately this doesn't work for me. It seems to download the source, but fails to compile

```
C:\dev\debug>npm install gulp-imagemin
-
> jpegtran-bin@3.0.2 postinstall C:\dev\debug\node_modules\gulp-imagemin\node_mo
dules\imagemin\node_modules\imagemin-jpegtran\node_modules\jpegtran-bin
> node lib/install.js

  ‼ Request to https://raw.github.com/imagemin/jpegtran-bin/3.0.2/vendor/win/x64
/libjpeg-62.dll failed
  ‼ jpegtran pre-build test failed
  i compiling from source
  × Error: ./configure --disable-shared --prefix="C:\dev\debug\node_modules\gulp
-imagemin\node_modules\imagemin\node_modules\imagemin-jpegtran\node_modules\jpeg
tran-bin\vendor" --bindir="C:\dev\debug\node_modules\gulp-imagemin\node_modules\
imagemin\node_modules\imagemin-jpegtran\node_modules\jpegtran-bin\vendor" && mak
e install
Command failed: C:\WINDOWS\system32\cmd.exe /s /c "./configure --disable-shared
...
```

Could the missing admin-rights be an issue?

**Comment by radum** [comment]
I'm behind a corporate proxy also, until recently the proxy was allowing anonymous connections until they decided to remove this "shitty" feature so everything broke. I tried changing the ENV vars and all, sometimes it worked sometimes it didn't. Also there is a huge security risk to have your password in plain txt in there, but even so there are so many things that can go wrong that the only solution I managed to find was to install FreeCap or ProxyCap and configure node.exe to be tunnelled by one of these 2 apps. I removed all proxy settings from .npmrc and ENV vars and now it works as expected.

The tool will handle the proxy stuff and will let the traffic flow to node as it was direct. All my other apps will be running normally without any interference from the tool because only node.exe is captured as I said.

Everyone has a different setting and until we burn all proxies to the ground it will never be fully solved. But we all know this will never happen in a corporate environment.

In the same time some package developers don't really care about proxies and said they will never do it until a good PR comes along, but even so it might not work.

So far I've been running node with FreeCap for 2 days now and had no issues yet.

**Comment by Rakeshcv** [comment]
Even I am experience same issue, can't download imagemin.

node version - 0.12.4 & npm version - 2.10.1

**Comment by alexpaluzzi** [comment]
Same problem, same three libraries. Also behind proxy. All other dependencies download without issue.

Anyone have luck working around this? (Without downgrading)

**Comment by Rakeshcv** [comment]
@alexpaluzzi what OS are you using? By any chance if u are using Mac, can you post the output of "openssl version -a"

**Comment by rdecastro** [comment]
I also have the same problem. The only workaround I found was to do npm install on a network that doesn't have any proxies. The same 3 libraries will install if you are outside of your work's firewall. ;(...

**Comment by alexpaluzzi** [comment]
@rdecastro see my resolution here:
https://github.com/sindresorhus/got/issues/79

**Comment by ac-astuartkregor** [comment]
We've been having problems with this as well. Eventually managed to get it to work by ensuring bin-build 2.2.1 was present before running npm install, however even though jpegtran is available on the path, the node package still insists on compiling it from source. Anyone experiencing the same symptoms?

**Comment by marc-guenther** [comment]
mikeal/tunnel-agent#15 fixed this problem for us.

**Comment by marc-guenther** [comment]
which has just been released as 0.4.2 :)
